src/Cli/dotnet/commands/dotnet-test/MSBuildHandler.cs

```diff
@@ -0,0 +1,231 @@
+        {
+            var projectCollection = new ProjectCollection();
+            var project = projectCollection.LoadProject(projectFilePath);
+            var buildResult = RestoreProject(projectFilePath, allowBinLog, projectCollection);
+
+            bool restored = buildResult.OverallResult == BuildResultCode.Success;
+
+            if (!restored)
+            {
+                return (Array.Empty<Module>(), restored);
+            }
+
+            return (ExtractModulesFromProject(project), restored);
+        }
+
+        private static IEnumerable<Module> ExtractModulesFromProject(Project project)
+        {
+            _ = bool.TryParse(project.GetPropertyValue(ProjectProperties.IsTestingPlatformApplication), out bool isTestingPlatformApplication);
+            _ = bool.TryParse(project.GetPropertyValue(ProjectProperties.IsTestProject), out bool isTestProject);
+
+            string targetFramework = project.GetPropertyValue(ProjectProperties.TargetFramework);
+            string targetFrameworks = project.GetPropertyValue(ProjectProperties.TargetFrameworks);
+            string targetPath = project.GetPropertyValue(ProjectProperties.TargetPath);
+            string projectFullPath = project.GetPropertyValue(ProjectProperties.ProjectFullPath);
+            string runSettingsFilePath = project.GetPropertyValue(ProjectProperties.RunSettingsFilePath);
+
+            var projects = new List<Module>();
+
+            if (string.IsNullOrEmpty(targetFrameworks))
+            {
+                projects.Add(new Module(targetPath, projectFullPath, targetFramework, runSettingsFilePath, isTestingPlatformApplication, isTestProject));
+            }
+            else
+            {
+                var frameworks = targetFrameworks.Split(Separator, StringSplitOptions.RemoveEmptyEntries);
+                foreach (var framework in frameworks)
+                {
+                    project.SetProperty(ProjectProperties.TargetFramework, framework);
+                    project.ReevaluateIfNecessary();
+
+                    projects.Add(new Module(project.GetPropertyValue(ProjectProperties.TargetPath),
+                        projectFullPath,
+                        framework,
+                        runSettingsFilePath,
+                        isTestingPlatformApplication,
+                        isTestProject));
+                }
+            }
+
+            return projects;
+        }
+
+        private static BuildResult RestoreProject(string projectFilePath, bool allowBinLog, ProjectCollection projectCollection)
+        {
+            BuildParameters parameters = new(projectCollection)
+            {
+                Loggers = [new ConsoleLogger(LoggerVerbosity.Quiet)]
+            };
+
+            if (allowBinLog)
+            {
+                parameters.Loggers = parameters.Loggers.Concat([
+                    new BinaryLogger
+                    {
+                        Parameters = BinLogFileName
```

Comment:
I think we should try to reuse some helper from the rest of SDK for that because there are ways to configure the binlog file name that we ignore here.